plugins/uefi/fu-uefi-pcrs.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-common.h"
+#include "fu-uefi-pcrs.h"
+
+typedef struct {
+	guint		 idx;
+	gchar		*checksum;
+} FuUefiPcrItem;
+
+struct _FuUefiPcrs {
+	GObject		 parent_instance;
+	GPtrArray	*items;		/* of FuUefiPcrItem */
+};
+
+G_DEFINE_TYPE (FuUefiPcrs, fu_uefi_pcrs, G_TYPE_OBJECT)
+
+static void
+fu_uefi_pcrs_parse_line (const gchar *line, gpointer user_data)
+{
+	FuUefiPcrs *self = FU_UEFI_PCRS (user_data);
+	FuUefiPcrItem *item;
+	guint64 idx;
+	g_autofree gchar *idxstr = NULL;
+	g_auto(GStrv) split = NULL;
+	g_autoptr(GString) str = NULL;
+
+	/* split into index:hash */
+	if (line == NULL || line[0] == '\0')
+		return;
+	split = g_strsplit (line, ":", 2);
+	if (g_strv_length (split) != 2) {
+		g_warning ("unexpected format, skipping: %s", line);
+		return;
+	}
+
+	/* get index */
+	idxstr = fu_common_strstrip (split[0]);
+	idx = fu_common_strtoull (idxstr);
+	if (idx > 64) {
+		g_warning ("unexpected index, skipping: %s", idxstr);
+		return;
+	}
+
+	/* parse hash */
+	str = g_string_new (split[1]);
+	if (str->len < 16)
+		return;
+	fu_common_string_replace (str, " ", "");
+	g_string_ascii_down (str);
+	item = g_new0 (FuUefiPcrItem, 1);
+	item->idx = idx;
+	item->checksum = g_string_free (g_steal_pointer (&str), FALSE);
+	g_ptr_array_add (self->items, item);
+	g_debug ("added PCR-%02u=%s", item->idx, item->checksum);
+}
+
+static gboolean
+fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error)
+{
+	g_auto(GStrv) lines = NULL;
+	g_autofree gchar *buf_pcrs = NULL;
+
+	/* get entire contents */
+	if (!g_file_get_contents (fn_pcrs, &buf_pcrs, NULL, error))
+		return FALSE;
+
+	/* find PCR lines */
+	lines = g_strsplit (buf_pcrs, "\n", -1);
+	for (guint i = 0; lines[i] != NULL; i++) {
+		if (g_str_has_prefix (lines[i], "PCR-"))
+			fu_uefi_pcrs_parse_line (lines[i] + 4, self);
+	}
+	return TRUE;
+}
+
+gboolean
+fu_uefi_pcrs_setup (FuUefiPcrs *self, GError **error)
+{
+	g_autofree gchar *devpath = NULL;
+	g_autofree gchar *sysfstpmdir = NULL;
+	g_autofree gchar *fn_pcrs = NULL;
+	const gchar *test_yaml = g_getenv ("FWUPD_UEFI_TPM2_YAML_DATA");
+
+	g_return_val_if_fail (FU_IS_UEFI_PCRS (self), FALSE);
+
+	/* check the TPM device exists at all */
+	sysfstpmdir = fu_common_get_path (FU_PATH_KIND_SYSFSDIR_TPM);
+	devpath = g_build_filename (sysfstpmdir, "tpm0", NULL);
+	if (!g_file_test (devpath, G_FILE_TEST_EXISTS)) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "no TPM device found");
+		return FALSE;
+	}
+
+	/* look for TPM 1.2 */
+	fn_pcrs = g_build_filename (devpath, "pcrs", NULL);
+	if (test_yaml == NULL && g_file_test (fn_pcrs, G_FILE_TEST_EXISTS))
+		return fu_uefi_pcrs_setup_tpm12 (self, fn_pcrs, error);
+
+	/* assume TPM 2.0 */
+	if (test_yaml == NULL) {
+		const gchar *argv[] = { "tpm2_pcrlist", NULL };
```

Comment:
helpfully I think that this tool renames in some version.
it's `tpm2_listpcrs` in some versions.